### PR TITLE
Don't import internally from types in `next-env.d.ts`

### LINKED ISSUE #29788: Cannot find name 'StaticImageData'.
### What version of Next.js are you using?

11.1.2 & 11.1.3-canary.57

### What version of Node.js are you using?

14.16.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next build

### Describe the Bug

The following error occurs during type checking.

```bash
node_modules/next/dist/client/image.d.ts:19:14 - error TS2304: Cannot find name 'StaticImageData'.

19     default: StaticImageData;
                ~~~~~~~~~~~~~~~

node_modules/next/dist/client/image.d.ts:21:45 - error TS2304: Cannot find name 'StaticImageData'.

21 declare type StaticImport = StaticRequire | StaticImageData;
                            
```

I got the following Issue in v11.0.0, upgraded to v11.1.2 and the problem was solved, but now I get this error instead.
https://github.com/vercel/next.js/issues/26892

### Expected Behavior

TS compiler should not give errors.

### To Reproduce

Here's the full code to my component:

```js
import { memo, useState, VFC } from 'react';
import Image, { ImageProps } from 'next/image';
import { isTest } from 'src/lib/environmentDetector';

type PropsType = ImageProps & {
  fallback: string | JSX.Element;
};

export const ImageWithFallback: VFC<PropsType> = memo(
  ({ src, fallback, ...rest }) => {
    const [showFallback, setShowFallback] = useState(false);
    if (isTest()) return null;

    const onError = () => {
      setShowFallback(true);
    };

    if (showFallback && typeof fallback !== 'string') return fallback;
    return <Image {...rest} src={showFallback ? (fallback as string) : src} onError={onError} />;
  },
  (p, n) => p.src === n.src
);
```

### PR DESCRIPTION
**Pull request by lfades**
Fixes https://github.com/vercel/next.js/issues/29788

This PR fixes an issue where `StaticImageData` is declared inside a global declaration file (`image-types/global.d.ts`) but it's not available in the published package.

Having `StaticImageData` defined globally works when working inside the Next.js repo because TypeScript takes care of adding global declarations to the scope, however the same isn't true when the package is published, the declaration file doesn't get included by default and therefore using the `StaticImageData` fails inside the image component types.

With this change the type is now defined in the Image component and the global declaration file is importing it from there instead, this file is also excluded in tsconfig to make sure it doesn't break when developing inside the Next.js repo, because the import goes to `/dist`

## Bug

- [x] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

### CODE REVIEW

**Review by styfle — DISMISSED** [review]
Nice catch, thanks!

**Review by ijjk — CHANGES_REQUESTED** [review]
Are we able to add a test case to prevent regressing on this? If the test needs to be run outside the Next.js repo we can use a [production test](https://github.com/vercel/next.js/tree/canary/test/production) which is isolated

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 88fe2544717a19ce36306ca22dc0c78bf5692c48

`yarn testheadless test/integration/image-component/svgo-webpack/test/index.test.js` 
- svgo-webpack with Image Component > next build > should not fail to build invalid usage of the Image component

<details>
<summary>Expand output</summary>

  ● svgo-webpack with Image Component › next build › should not fail to build invalid usage of the Image component

    expect(received).toBeFalsy()

    Received: "Failed to compile.·
    ./pages/index.tsx:1:18
    Type error: Cannot find module '../public/test.svg' or its corresponding type declarations.·
    > 1 | import Icon from '../public/test.svg'
        |                  ^
      2 | 
      3 | export default function Home() {
      4 |   return <Icon />
    "

      19 |     it('should not fail to build invalid usage of the Image component', async () => {
      20 |       const { stderr, code } = await nextBuild(appDir, [], { stderr: true })
    > 21 |       expect(stderr).toBeFalsy()
         |                      ^
      22 |       expect(code).toBe(0)
      23 |     })
      24 |   })

      at Object.<anonymous> (integration/image-component/svgo-webpack/test/index.test.js:21:22)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| buildDuration | 19.3s | 18.4s | -902ms |
| buildDurationCached | 7.1s | 7.1s | ⚠️ +51ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +97 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.072 | 4.097 | ⚠️ +0.03 |
| / avg req/sec | 613.94 | 610.19 | ⚠️ -3.75 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.999 | 1.995 | 0 |
| /error-in-render avg req/sec | 1250.5 | 1252.83 | +2.33 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 5.01 kB | 5.01 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| buildDuration | 23.2s | 23.4s | ⚠️ +254ms |
| buildDurationCached | 7.1s | 7.4s | ⚠️ +251ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +97 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.043 | 4.134 | ⚠️ +0.09 |
| / avg req/sec | 618.37 | 604.78 | ⚠️ -13.59 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.996 | 2.083 | ⚠️ +0.09 |
| /error-in-render avg req/sec | 1252.26 | 1200.25 | ⚠️ -52.01 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.1 kB | 42.1 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| image-HASH.js gzip | 5.05 kB | 5.05 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.28 kB | 2.28 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 548 B | 548 B | ✓ |
| withRouter.html gzip | 529 B | 529 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: 88fe2544717a19ce36306ca22dc0c78bf5692c48

**Comment by lfades** [comment]
@ijjk If the change is added in the future again it will be caught by our type checking before publishing next, because it's now excluded in tsconfig and so trying to import from there won't work, this is the same behavior as the `*.d.ts` files we have in the root.

I tried doing the same for `types/global.d.ts` but we're internally relying in the namespace changes we do to `NodeJS`, however there is no issue here because even though we rely on it, we don't rely on exported interfaces/types from it, so the generated declaration files don't care about it, unless of course an interface/type is added there and imported somewhere within the source, that's unlikely however, we can add a note, wdyt?

**Comment by ijjk** [comment]
## Failing test suites 
Commit: a6067defc0ff585daf1163934199649b07cf9428

`yarn testheadless test/integration/image-component/svgo-webpack/test/index.test.js` 
- svgo-webpack with Image Component > next build > should not fail to build invalid usage of the Image component

<details>
<summary>Expand output</summary>

  ● svgo-webpack with Image Component › next build › should not fail to build invalid usage of the Image component

    expect(received).toBeFalsy()

    Received: "Failed to compile.·
    ./pages/index.tsx:1:18
    Type error: Cannot find module '../public/test.svg' or its corresponding type declarations.·
    > 1 | import Icon from '../public/test.svg'
        |                  ^
      2 | 
      3 | export default function Home() {
      4 |   return <Icon />
    "

      19 |     it('should not fail to build invalid usage of the Image component', async () => {
      20 |       const { stderr, code } = await nextBuild(appDir, [], { stderr: true })
    > 21 |       expect(stderr).toBeFalsy()
         |                      ^
      22 |       expect(code).toBe(0)
      23 |     })
      24 |   })

      at Object.<anonymous> (integration/image-component/svgo-webpack/test/index.test.js:21:22)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| buildDuration | 19.2s | 19.4s | ⚠️ +186ms |
| buildDurationCached | 7.6s | 7.5s | -95ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +79 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.959 | 4.037 | ⚠️ +0.08 |
| / avg req/sec | 631.42 | 619.32 | ⚠️ -12.1 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.838 | 1.787 | -0.05 |
| /error-in-render avg req/sec | 1359.81 | 1398.97 | +39.16 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 5.01 kB | 5.01 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| buildDuration | 23.1s | 23.1s | -15ms |
| buildDurationCached | 7.6s | 7.4s | -175ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +79 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.859 | 3.919 | ⚠️ +0.06 |
| / avg req/sec | 647.9 | 637.91 | ⚠️ -9.99 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.697 | 1.733 | ⚠️ +0.04 |
| /error-in-render avg req/sec | 1473.31 | 1442.23 | ⚠️ -31.08 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.1 kB | 42.1 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| image-HASH.js gzip | 5.05 kB | 5.05 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.28 kB | 2.28 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lfades/next.js fix/unexpected-type-import | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 548 B | 548 B | ✓ |
| withRouter.html gzip | 529 B | 529 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: a6067defc0ff585daf1163934199649b07cf9428

File: packages/next/image-types/global.d.ts [hunk 1]
```diff
@@ -1,15 +1,8 @@
...
-interface StaticImageData {
-  src: string
-  height: number
-  width: number
-  blurDataURL?: string
-}
-
 declare module '*.png' {
-  const content: StaticImageData
+  const content: import('../dist/client/image').StaticImageData
```
Comment (lfades) [inline packages/next/image-types/global.d.ts:5]: TIL if I add a root import in this file TS will no longer treat it as a global declaration file, but using import within the ambient declaration works

**Review by styfle — APPROVED** [review]

**Pull request merged**